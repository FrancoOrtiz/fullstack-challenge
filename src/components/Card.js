import React from 'react'

export const Card = () => {
  return (
    <div className="flex w-80 h-40 justify-between my-10">
      <div className="h-full w-40 bg-gray-500">
        <img src=""></img>
      </div>
      <span className="mt-2 text-xl text-gray-500">Pikachu</span>
    </div>
  )
}
