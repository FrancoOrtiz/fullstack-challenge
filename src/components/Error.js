import React from "react";

export const Error = ({ message }) => {
  return (
    <div className="bg-red-200 border-2 border-red-500 rounded-lg flex items-center justify-center text-red-600 m-auto mb-4 w-full h-14">
      {message}
    </div>
  );
};
