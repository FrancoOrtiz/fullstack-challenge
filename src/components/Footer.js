import React from "react";

export const Footer = () => {
  return (
    <footer className="w-4/5 m-auto flex justify-between mt-10 border-t-2 border-gray-400 items-center h-20 pl-10 pr-20">
      <p>Hecho por Franco</p>
      <a
        href="https://github.com/FrancoOrtiz/"
        target="blank"
        className="w-1/4 py-2 shadow-button text-center text-xl rounded-xl bg-gray-300 font-bold border-black border-2"
      >
        Link a mi repo
      </a>
    </footer>
  );
};
