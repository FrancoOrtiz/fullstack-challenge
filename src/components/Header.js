import React from "react";

export const Header = () => {
  return (
    <header className="flex flex-col gap-3 w-4/5 m-auto">
      <h1 className="text-black text-6xl font-bold">Pokemon finder</h1>
      <p className="text-gray-500 text-xl">
        El que quiere Pokemons, que los busque.
      </p>
    </header>
  );
};
