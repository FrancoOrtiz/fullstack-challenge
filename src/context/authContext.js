import { createContext, useContext, useEffect, useState } from "react";
import {
  createUserWithEmailAndPassword,
  signInWithEmailAndPassword,
  onAuthStateChanged,
  signOut,
  GoogleAuthProvider,
  signInWithPopup,
  sendPasswordResetEmail,
} from "firebase/auth";
import { auth, db } from "../firebase";
import { setDoc, doc, getDoc, onSnapshot } from "firebase/firestore";

export const authContext = createContext();

export const useAuth = () => {
  const context = useContext(authContext);
  return context;
};

export const AuthProvider = ({ children }) => {
  const [user, setUser] = useState(null);
  const [loading, setLoading] = useState(true);
  const [favPokemon, setFavPokemon] = useState([]);

  const getFavPokemons = async (currentUser) => {
    if (currentUser) {
      const docRef = doc(db, "users", currentUser.uid);
      const docSnap = await getDoc(docRef);
      const { favorites } = docSnap.data();
      if (favorites) {
        setFavPokemon(favorites);
      } else {
        setFavPokemon([]);
      }
    }
  };

  const signup = (email, password) => {
    return createUserWithEmailAndPassword(auth, email, password);
  };

  const login = (email, password) => {
    return signInWithEmailAndPassword(auth, email, password);
  };

  const logout = () => signOut(auth);

  const addFavPokemon = async (name) => {
    const actualFav = favPokemon.find((fav) => fav === name);
    const indexPokemon = favPokemon.indexOf(name);
    const docRef = doc(db, "users", user.uid);

    if (actualFav) {
      favPokemon.splice(indexPokemon, 1);
    } else {
      favPokemon.push(name);
    }

    await setDoc(docRef, { favorites: favPokemon });
    await onSnapshot(docRef, (doc) => {
      const { favorites } = doc.data();
      setFavPokemon(favorites);
    });
  };

  const loginWithGoogle = () => {
    const googleProvider = new GoogleAuthProvider();
    return signInWithPopup(auth, googleProvider);
  };

  const resetPassword = (email) => sendPasswordResetEmail(auth, email);

  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, (currentUser) => {
      setUser(currentUser);
      setLoading(false);
      getFavPokemons(currentUser);
    });

    return () => unsubscribe();
  }, []);

  return (
    <authContext.Provider
      value={{
        signup,
        login,
        user,
        logout,
        loginWithGoogle,
        loading,
        resetPassword,
        addFavPokemon,
        favPokemon,
      }}
    >
      {children}
    </authContext.Provider>
  );
};
