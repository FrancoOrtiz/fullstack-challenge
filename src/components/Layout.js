import React, {Fragment} from 'react'
import { Footer } from './Footer'
import { Header } from './Header'

export const Layout = ({children}) => {
  return (
    <Fragment>
        <Header/>
            {children}
        <Footer/>
    </Fragment>
  )
}
