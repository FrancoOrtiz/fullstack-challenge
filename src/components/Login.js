import React, { useState } from "react";
import { useNavigate } from "react-router-dom";
import { useAuth } from "../context/authContext";
import { Header } from "./Header";
import { Error } from "./Error";

export const Login = () => {
  const [user, setUser] = useState({
    email: "",
    password: "",
  });
  const [error, setError] = useState();
  const { login, loginWithGoogle, resetPassword } = useAuth();
  const navigate = useNavigate();

  const handleChange = ({ target: { name, value } }) => {
    setUser({ ...user, [name]: value });
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    try {
      await login(user.email, user.password);
      navigate("/");
    } catch (error) {
      setError(error.message);
    }
  };

  const handleGoogleLogin = async () => {
    try {
      await loginWithGoogle();
      navigate("/");
    } catch (error) {
      setError(error.code);
    }
  };

  const handleForgot = async () => {
    if (!user.email) {
      setError("Ingrese su correo");
    } else {
      try {
        await resetPassword(user.email);
        setError("Te hemos enviado un correo para que recuperes tu contraseña");
      } catch (error) {
        setError(error.message);
      }
    }
  };

  const navRegister = () => {
    navigate("/register");
  };

  return (
    <div className="mt-32 w-4/5 lg:w-2/4 m-auto">
      {error && <Error message={error} />}

      <form
        onSubmit={handleSubmit}
        className=" pt-10 pb-6 flex flex-col text-center gap-6 bg-white rounded-xl shadow-lg shadow-slate-400"
      >
        <Header></Header>
        <div className="w-4/5 m-auto flex flex-col gap-1 my-0">
          <label htmlFor="email">Email</label>
          <input
            type="email"
            name="email"
            placeholder="youremail@company.ltd"
            className="border-solid border-2 rounded-md border-gray-500 h-10 pl-2 w-2/3 lg:w-1/2 m-auto"
            onChange={handleChange}
          />
        </div>

        <div className="w-4/5 m-auto flex flex-col gap-1 my-0">
          <label htmlFor="password">Password</label>
          <input
            type="password"
            name="password"
            id="password"
            placeholder="******"
            className="border-solid border-2 rounded-md border-gray-500 h-10 pl-2 w-2/3 lg:w-1/2 m-auto"
            onChange={handleChange}
          />
        </div>

        <div className="w-full flex-col items-center flex pt-6 mt-6 border-solid border-t-2 border-gray-200">
          <button className="w-44 h-14 bg-blue-500 font-bold text-white rounded-lg text-lg hover:bg-transparent border-2 border-blue-500 hover:text-blue-500 transition-all">
            Login
          </button>
          <button
            type="button"
            onClick={handleGoogleLogin}
            className="w-52 h-14 border-2 font-bold border-black rounded-lg text-lg mt-4 "
          >
            Login with Google
          </button>
        </div>
      </form>
      <div className="mt-4 flex w-full justify-between">
        <button
          type="button"
          className="text-blue-400 mt-4 hover:text-blue-600"
          onClick={navRegister}
        >
          No tienes un usuario? Clickea aqui
        </button>
        <button
          type="button"
          className="text-blue-400 mt-4 hover:text-blue-600"
          onClick={handleForgot}
        >
          Olvidaste tu contraseña?
        </button>
      </div>
    </div>
  );
};
