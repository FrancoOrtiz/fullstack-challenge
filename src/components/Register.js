import React, { useState } from "react";
import { useNavigate } from "react-router-dom";
import { useAuth } from "../context/authContext";
import { Error } from "./Error";
import { Header } from "./Header";

export const Register = () => {
  const [user, setUser] = useState({
    email: "",
    password: "",
  });
  const [error, setError] = useState();
  const { signup } = useAuth();
  const navigate = useNavigate();

  const handleChange = ({ target: { name, value } }) => {
    setUser({ ...user, [name]: value });
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    try {
      await signup(user.email, user.password);
      navigate("/");
    } catch (error) {
      setError(error.message);
    }
  };
  const navLogin = () => {
    navigate("/login");
  };

  return (
    <div className="mt-32 w-4/5 lg:w-2/4 m-auto">
      {error && <Error message={error} />}

      <form
        onSubmit={handleSubmit}
        className=" pt-10 pb-6 flex flex-col text-center gap-6 bg-white rounded-xl shadow-lg shadow-slate-400"
      >
        <Header></Header>
        <div className="w-4/5 m-auto flex flex-col gap-1 my-0">
          <label htmlFor="email">Email</label>
          <input
            type="email"
            name="email"
            placeholder="youremail@company.ltd"
            className="border-solid border-2 rounded-md border-gray-500 h-10 pl-2 w-2/3 lg:w-1/2 m-auto"
            onChange={handleChange}
          />
        </div>

        <div className="w-4/5 m-auto flex flex-col gap-1 my-0">
          <label htmlFor="password">Password</label>
          <input
            type="password"
            name="password"
            id="password"
            placeholder="******"
            className="border-solid border-2 rounded-md border-gray-500 h-10 pl-2 w-2/3 lg:w-1/2 m-auto"
            onChange={handleChange}
          />
        </div>

        <div className="w-full flex-col items-center flex pt-6 mt-6 border-solid border-t-2 border-gray-200">
          <button className="w-44 h-14 font-bold bg-blue-500 text-white rounded-lg text-lg hover:bg-transparent border-2 border-blue-500 hover:text-blue-500 transition-all">
            Register
          </button>
        </div>
      </form>
      <button
        type="button"
        className="text-blue-400 mt-6 hover:text-blue-600"
        onClick={navLogin}
      >
        Ya tienes un usuario? Clickea aqui
      </button>
    </div>
  );
};
