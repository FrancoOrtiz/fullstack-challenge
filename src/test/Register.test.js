import { Register } from "components/Register";
import { fireEvent, render, waitFor } from "@testing-library/react";
import { authContext } from "context/authContext";
import React from "react";
import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";

describe("Tests on Register.js", () => {
  let wrapper;
  const mockSignup = jest.fn();

  test("should be render", () => {
    wrapper = renderer
      .create(
        <authContext.Provider value={{}}>
          <BrowserRouter>
            <Register />
          </BrowserRouter>
        </authContext.Provider>
      )
      .toJSON();

    expect(wrapper).toMatchSnapshot();
  });

  test("should be call signup function", async () => {
    wrapper = render(
      <authContext.Provider value={{ signup: mockSignup }}>
        <BrowserRouter>
          <Register />
        </BrowserRouter>
      </authContext.Provider>
    );
    const btn = wrapper.getByText("Register");
    expect(btn).toBeDefined();

    waitFor(() => fireEvent.click(btn));
    expect(mockSignup).toBeCalledTimes(1);
  });

  test("should be called with the arguments email and password", () => {
    wrapper = render(
      <authContext.Provider value={{ signup: mockSignup }}>
        <BrowserRouter>
          <Register />
        </BrowserRouter>
      </authContext.Provider>
    );

    const email = wrapper.getByPlaceholderText("youremail@company.ltd");
    fireEvent.change(email, { target: { value: "tests@email.com" } });
    const password = wrapper.getByPlaceholderText("******");
    fireEvent.change(password, { target: { value: "123456" } });

    const btn = wrapper.getByText("Register");
    expect(btn).toBeDefined();

    waitFor(() => fireEvent.click(btn));
    expect(mockSignup).toHaveBeenCalledWith("tests@email.com", "123456");
  });
});
