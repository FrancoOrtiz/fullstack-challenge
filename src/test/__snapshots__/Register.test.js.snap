// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on Register.js should be render 1`] = `
<div
  className="mt-32 w-4/5 lg:w-2/4 m-auto"
>
  <form
    className=" pt-10 pb-6 flex flex-col text-center gap-6 bg-white rounded-xl shadow-lg shadow-slate-400"
    onSubmit={[Function]}
  >
    <header
      className="flex flex-col gap-3 w-4/5 m-auto"
    >
      <h1
        className="text-black text-6xl font-bold"
      >
        Pokemon finder
      </h1>
      <p
        className="text-gray-500 text-xl"
      >
        El que quiere Pokemons, que los busque.
      </p>
    </header>
    <div
      className="w-4/5 m-auto flex flex-col gap-1 my-0"
    >
      <label
        htmlFor="email"
      >
        Email
      </label>
      <input
        className="border-solid border-2 rounded-md border-gray-500 h-10 pl-2 w-2/3 lg:w-1/2 m-auto"
        name="email"
        onChange={[Function]}
        placeholder="youremail@company.ltd"
        type="email"
      />
    </div>
    <div
      className="w-4/5 m-auto flex flex-col gap-1 my-0"
    >
      <label
        htmlFor="password"
      >
        Password
      </label>
      <input
        className="border-solid border-2 rounded-md border-gray-500 h-10 pl-2 w-2/3 lg:w-1/2 m-auto"
        id="password"
        name="password"
        onChange={[Function]}
        placeholder="******"
        type="password"
      />
    </div>
    <div
      className="w-full flex-col items-center flex pt-6 mt-6 border-solid border-t-2 border-gray-200"
    >
      <button
        className="w-44 h-14 font-bold bg-blue-500 text-white rounded-lg text-lg hover:bg-transparent border-2 border-blue-500 hover:text-blue-500 transition-all"
      >
        Register
      </button>
    </div>
  </form>
  <button
    className="text-blue-400 mt-6 hover:text-blue-600"
    onClick={[Function]}
    type="button"
  >
    Ya tienes un usuario? Clickea aqui
  </button>
</div>
`;
